File: Server_Code/Packet/LoginModelSelectReqPacket.cpp
#include "LoginModelSelectReqPacket.h"

LoginModelSelectReqPacket::LoginModelSelectReqPacket(uint8_t modelId)
    :LoginPacket(EPacketType::LOGIN_MODEL_SELECT_REQ), modelId(modelId)
{
}

std::vector<char> LoginModelSelectReqPacket::Serialize() const
{
    BinaryWriter writer;
    writer.WriteHeader(header);
    writer.WriteUInt8(modelId);
    return writer.GetBuffer();
}

void LoginModelSelectReqPacket::Deserialize(const char* buffer, size_t bufferSize)
{
    BinaryReader reader(buffer, bufferSize);
    header = reader.ReadHeader();
    modelId = reader.ReadUInt8();

    if (modelId > MAX_MODEL_ID) {
        throw std::invalid_argument("Invalid model ID selected");
    }
}

File: Assets/Script/Packet/LoginModelSelectReqPacket.cs
using System;
using System.Text;
using NetworkUtil;

public class LoginModelSelectReqPacket : LoginPacket
{
    public override EPacketType PacketType => EPacketType.LOGIN_MODEL_SELECT_REQ;

    public byte ModelId { get; private set; }

    public LoginModelSelectReqPacket() { }

    public LoginModelSelectReqPacket(byte modelId, string nickname) : base(nickname)
    {
        ModelId = modelId;
    }

    public override void Serialize(BinaryWriter writer)
    {
        writer.WriteUInt8(ModelId);
        writer.WriteString(Nickname); // UTF-8, 길이 포함
    }

    public override void Deserialize(BinaryReader reader)
    {
        ModelId = reader.ReadUInt8();
        if (ModelId > 3)
            throw new ArgumentException("Invalid model ID");

        Nickname = reader.ReadString();
    }

    public override bool IsValid()
    {
        return ModelId <= 3 && !string.IsNullOrEmpty(Nickname) && Encoding.UTF8.GetByteCount(Nickname) <= 255;
    }
}
